### ISSUE TITLE
Asyncio Tutorial

### ISSUE THREAD

**Issue by 60717714-ac02-4c78-a5e0-6f0d36b9c456** [python/cpython#79012]
BPO | [34831](https://bugs.python.org/issue34831)
--- | :---
Nosy | @asvetlov, @cjrh, @elprans, @1st1, @willingc
PRs | <li>python/cpython#9748</li>

<sup>*Note: these values reflect the state of the issue at the time it was migrated and might not reflect the current state.*</sup>

<details><summary>Show more details</summary><p>

GitHub fields:
```python
assignee = None
closed_at = None
created_at = <Date 2018-09-28.11:36:17.041>
labels = ['3.8', 'type-feature', '3.7', 'docs', 'expert-asyncio']
title = 'Asyncio Tutorial'
updated_at = <Date 2019-06-19.12:39:06.198>
user = 'https://github.com/cjrh'
```

bugs.python.org fields:
```python
activity = <Date 2019-06-19.12:39:06.198>
actor = 'cjrh'
assignee = 'docs@python'
closed = False
closed_date = None
closer = None
components = ['Documentation', 'asyncio']
creation = <Date 2018-09-28.11:36:17.041>
creator = 'cjrh'
dependencies = []
files = []
hgrepos = []
issue_num = 34831
keywords = ['patch']
message_count = 14.0
messages = ['326628', '326665', '326789', '326977', '327266', '327271', '327275', '327284', '327286', '328201', '333088', '345661', '345755', '346039']
nosy_count = 6.0
nosy_names = ['asvetlov', 'cjrh', 'Elvis.Pranskevichus', 'docs@python', 'yselivanov', 'willingc']
pr_nums = ['9748']
priority = 'normal'
resolution = None
stage = 'patch review'
status = 'open'
superseder = None
type = 'enhancement'
url = 'https://bugs.python.org/issue34831'
versions = ['Python 3.7', 'Python 3.8']
```

</p></details>

### Linked PRs
* gh-144594
* gh-144604

**Comment by cjrh** [comment]
So, I basically stalled on this work because I was waiting for changes in asyncio to settle down. In particular task groups and PEP 654 exception groups. After PEP 654 in 3.11, there will be more changes to asyncio to make use of it yes?

**Comment by cjrh** [comment]
Ok I see they're already in #90908

**Comment by gvanrossum** [comment]
I doubt that asyncio will ever "settle down". :-) But anyway, I am currently working on [docs for task groups](https://github.com/python/cpython/pull/94359).

**Comment by kumaraditya303** [comment]
The PR https://github.com/python/cpython/pull/9748 seems stuck and the PR is very large, it would be better to split it up so that things get merged as ready.

@CAM-Gerlach Are you interested in taking over this? I and hopefully @gvanrossum can volunteer to review the docs updates. It would be great to have this in 3.12.

**Comment by cjrh** [comment]
Anyone is more than welcome to take it over. Unfortunately I haven't been able to find bandwidth to pick this up again.

**Comment by CAM-Gerlach** [comment]
This looks like a pretty big project, and as a disclaimer I have very little subject matter expertise or experience (coming from scientific Python, when I work with parallelism is almost invariably of the "embarrassingly parallel" compute-bound variety, and consequently use of `asyncio` is virtually unheard of in most of my discipline). Like with `sqlite3`, the other individual module I've spent a relatively disproportionate amount of time helping with the docs on, its something where most of what I do know comes from reviewing people's docs PRs on the subject (since at least in my domains of research, scientists tend to have a strong aversion to SQL-type databases as a data storage mechanism).

This can actually be a significant benefit to helping ensure the docs, especially a generally beginner-focused tutorial, is easy to understand and follow, and doesn't embed assumptions that can be really easy to forget about as a subject matter expert (as I do myself when I am in those shoes). However, for the parts that aren't already written, I'm going to need knowledge help actually drafting something, at which point I can learn the material myself and then make it better (which is typically what I am strongest at anyway). The sprints next week might be a great time to work on it if others have some bandwidth to help guide me and answer my inevitable questions, assuming I am not called away to more immediately pressing tasks.

My initial high level comments is based on the outline above, as well as skimming through the PR. Right now, it looks like much if not most of this content, aside from the "case study", is actually not quite a tutorial at all, at least as currently framed:

* Most of the "Why" section sounds a lot more like archetypal "explanation"-type content that is oriented toward understanding and serves an important need, but is abstract rather than concrete ("why" rather than "how") and separate from the goal of getting a newcomer jump into _using_ `asyncio`.
* The "def vs async def" section sounds mostly like reference material describing the differences between the two types of functions, perhaps mixed with some explanation in parts, rather than guiding a newcomer through a series of steps to complete a learning task
* "How to" run async def functions, and "dealing with concurrent tasks" could potentially be incorporated into a tutorial assuming it is basic enough and appropriate to the tutorial's topic, but as framed as a top-level section, it is, well, a how-to, not a tutorial, as it focuses on telling the user step by step how to solve specific problems.
* The Case Study, however, sounds like a solid basis for a great tutorial—it sounds fun, engaging, achievable by a beginner but also realistic and a significant accomplishment that likely exercises a good amount of asyncio's functionality. Furthermore, it seems its primary focus is on helping the user learn the basics of using asyncio through walking them through an appropriate example project.

Following this lens, each of these can be independent pages, or at least section of existing pages, and organized appropriately with other content that has a similar goal (and, ideally, in the order readers tend to refer to them, i.e. essentially the inverse of what they are now—i.e. something like Tutorial, How-To, Reference, Explanation (rather than the current opposite). However, that provides a great framework for splitting up the existing PR into more manageable and independent chunks that we could complete, revise and merge in series or in parallel, rather than a single large, interdependent set of documents. Its also a great opportunity for us to continue gaining more practical experience applying the Diataxis principles, as part of a broader process of of expanding and revising our documentation to better serve users' needs.

**Comment by gvanrossum** [comment]
I'm happy to answer all sorts of questions, both in person during the sprint and online afterwards.

I can't tell from skimming this issue how much was already written. Assuming not much has been written, my recommendation would be to start small, trying to get a PR merged that adds a very skinny tutorial that doesn't cover most topics (even important ones). That way we can hopefully spread the effort, perhaps similarly to how the Python tutorial grew from utter basics to what it is now, through many contributions.

I hope we can make this a showcase for Diataxis (I hope you did the workshop, I didn't, alas, but I've skimmed the docs about it).

**Comment by CAM-Gerlach** [comment]
> I'm happy to answer all sorts of questions, both in person during the sprint and online afterwards.

You've been really helpful answering my questions so far (including those I didn't know ask, but appreciated having answered) so I'm happy to have you on board!

> I can't tell from skimming this issue how much was already written. 

It wasn't formally linked here, but it seems from the corresponding PR, #9748 , that the great majority it was already written, except for the chat client CLI tutorial (aside from just a stub), and a few specific how-tos in the "cookbook" which were marked as TODO.

> Assuming not much has been written, my recommendation would be to start small, trying to get a PR merged that adds a very skinny tutorial that doesn't cover most topics (even important ones).

Since most of it is already written, what I had been thinking might work best pulling out each atomic top-level component of that, doing the necessary work to get it ready to go and submitting each for review as an independent PR (either as its own document, or as a section of an existing one as appropriate). We could simply defer the client tutorial for now as well as the still-TODO how-to items, at least until after this was all completed. Right now, the existing PR is fairly well enough organized into components that it shouldn't be too difficult to pull apart, with a bit of thought ahead of time and appropriate revisions.

> I hope we can make this a showcase for Diataxis 

Indeed, I hope so too—though, our primary goal is to serve users, with Diataxis being an effective means to that end.

> I hope you did the workshop

I'd hope so too, considering I was one of the organizers :laughing: and I also have some (very helpful) practical experience applying it alongside Ezio and Erlend to the `sqlite3` module docs (with the tutorial being what we ended up having to spend the most time on, and making the most improvements too).

> I've skimmed the docs about it

The docs are good on their own, but as I found out along with Ezio and Erlend, we only started to truly understand and appreciate Diataxis once we dug in and really started putting it into practice.

**Comment by CAM-Gerlach** [comment]
Unfortunately, I didn't get the chance to make much progress on this at the sprint as I was very bogged down on other things. I do have a fair number of urgent things to get out of the way first, but hopefully this should go fairly well once I get to it, since most of the work is already done in the PR, with the most of the remainder being right within my wheelhouse.

**Comment by kumaraditya303** [comment]
We should also consider adding some async generators examples in tutorial with current best practices. 

See also https://github.com/python/cpython/issues/100108

**Comment by willingc** [comment]
Changing the status of this issue to TODO. The original PR was closed. Thanks @cjrh for your work on that (:wave: hope all is well). This would be a good in-person effort at a future sprint.

**Comment by willingc** [comment]
There is current interest in #124668 for additional examples. See https://github.com/python/cpython/issues/124668#issuecomment-2418943439 for additional context.

**Comment by picnixz** [comment]
I kept the 3.12, 3.13 and 3.14 labels so that the tutorial is also backported. It doesn't make sense not to backport such tutorial if one is made (but the backports shoud be carefully done since they could be different depending on the available features).

**Comment by gvanrossum** [comment]
> I kept the 3.12, 3.13 and 3.14 labels so that the tutorial is also backported. It doesn't make sense not to backport such tutorial if one is made (but the backports shoud be carefully done since they could be different depending on the available features).

That's great aspirational thinking, though I fear that in practice we won't find volunteers to do the work for the backports, so the realistic choice is either no backport or an unreliable backport.

Maybe the "backports" could just hard-link to the 3.14 tutorial, and the latter could have a light sprinkling of "new in 3.13" and "new in 3.14" notes (but not "new in 3.12" or earlier versions).

**Comment by picnixz** [comment]
> Maybe the "backports" could just hard-link to the 3.14 tutorial

That could be a solution but I don't know how to do it without hardcoding the link. Now, since the issue has been opened for a long time, I think that 3.14 would likely be the "main" at the point it is written, and then we wouldn't have backports issues. Now, most of the stuff the asyncio tutorial would contain is likely to be compatible for 3.12+ so maybe the backports will be fine?

Nonetheless, I just wanted to keep the labels for myself and so that we can remember to address this when writing the tutorial (by the way, I would probably be fine with helping doing the backports, one possible issue I see is that it could create conflicts for subsequent PRs if the wording changes too much).

**Comment by gvanrossum** [comment]
Totally fine. We need the tutorial first, the backport can be decided later.

And 3.14 already *is* main.

**Comment by StanFromIreland** [comment]
cc @willingc 

The new Asyncio HOWTO has been added in https://github.com/python/cpython/commit/3964f974894eff1653913dda437971e0bbfa8ccc, I think that adding a tutorial would be quite duplicative? Can this be closed?

**Comment by AA-Turner** [comment]
Tutorials and explainers (HOWTO) fall under different sections of the [Diátaxis](https://diataxis.fr/) framework. It would still be useful to have a tutorial.

A

**Comment by ZeroIntensity** [comment]
> Tutorials and explainers (HOWTO) fall under different sections of the [Diátaxis](https://diataxis.fr/) framework. It would still be useful to have a tutorial.

Sorry, I disagree. "A Conceptual Overview of asyncio" functions as an asyncio tutorial in its current form. I think it's just badly named a "HOWTO" because there's no final goal of creating anything; it is merely something to help beginners grasp asyncio, which corresponds to the Diátaxis definition of a tutorial.

More generally, I don't think our definition of "HOWTO" fits Diátaxis. For example, the [free-threading page](https://docs.python.org/3/howto/free-threading-python.html#freethreading-python-howto), the [`ipaddress` introduction](https://docs.python.org/3/howto/ipaddress.html#ipaddress-howto), and the [descriptor guide](https://docs.python.org/3/howto/descriptor.html#descriptorhowto) are all marked as "HOWTO" articles, despite them not "guiding the user towards an action", per Diátaxis's definition of HOWTO.

I'm going to close this for the time being, since this issue inspired a [PR](https://github.com/python/cpython/pull/144594) that was very similar to the tutorial added in #137215.
